lib/RestructureCFG/SelectScopePass.cpp

```diff
@@ -0,0 +1,261 @@
+              "Nodes between conditional and its postdominator, in reverse "
+              "post order:\n");
+    for (auto DFSNode : NodesToProcess) {
+      revng_log(SelectScopePassLogger, "  " << DFSNode->getName().str());
+    }
+
+    // Initialize the `ReachabilityMap`
+    std::map<BasicBlock *, size_t> ReachabilityMap;
+
+    // We initialize the `ReachabilityMap` for each `ConditionalSuccessor`
+    for (const auto &[Index, ConditionalSuccessor] :
+         enumerate(ConditionalSuccessors)) {
+      if (not ReachabilityMap.contains(ConditionalSuccessor)) {
+        ReachabilityMap[ConditionalSuccessor] = Index;
+      }
+    }
+
+    // Process each node in the zone of interest
+    for (BasicBlock *Candidate : NodesToProcess) {
+      revng_log(SelectScopePassLogger,
+                "Analyzing candidate: " + Candidate->getName().str() << "\n");
+
+      // We precompute the predecessors to avoid invalidation due to graph
+      // changes. It is fundamental that we always traverse the `ScopeGraph`
+      // view of the CFG, or we may end up with some inconsistencies in terms
+      // of the visited nodes.
+      revng_log(SelectScopePassLogger, "The candidate predecessors are:\n");
+      SetVector<BasicBlock *> Predecessors = getPredecessors(Candidate);
+
+      // `Candidate`, could be itself a immediate successor of a conditional
+      // node, and therefore correspond to a `ScopeID`. We therefore need to
+      // take into consideration it when assigning the final scope for each
+      // `Candidate`.
+      // We can do this by always enqueuing `Candidate` as a predecessor of
+      // itself, this can lead to two situations:
+      // 1) `Candidate` is not a successor of the conditional, therefore no
+      //    corresponding entry in `ReachabilityMap` will be present, and this
+      //    will not influence the decision on the `ScopeID` which will be
+      //    finally assigned.
+      // 2) `Candidate` is a successor of the conditional, therefore a
+      //    corresponding entry in `ReachabilityMap` will be present, and it
+      //    will be correctly taken into account for the `ScopeID` decision
+      //    process.
+      // Alternatively, we could pre-assign the `ScopeID`, in the
+      // `ReachabilityMap`, for each successor of a conditional node during
+      // the initialization. This, however, would tie us to the decision of
+      // always assigning the successor of a conditional node to the `ScopeID`
+      // opening in the successor itself, while, in principle, we could
+      // alternatively disconnect the edge connecting the conditional and the
+      // successor, by making it a `goto` edge.
+      // Predecessors.push_back(Candidate);
+      Predecessors.insert(Candidate);
+
+      // TODO: we elect the first `ScopeID` that we encounter as the elected
+      //       `ScopeID`. We may employ a more complex strategy here.
+      std::optional<size_t> ElectedScopeID;
```

Comment:
Always picking the first seems very naive.

I thought at some point we said: look at all predecessors with a decided scope and pick the most common scope among the predecessors.

Reply:
This is causing some strange bugs down the line with IDS that are taking some time to investigate, so I'll leave this open until I solve

Reply:
Ok, I got to the bottom of this, nothing strange just I didn't pay attention to a couple of details.

The election is now performed with the method you mentioned, I wrote a small routine doing this here https://github.com/revng/revng/blob/f6e301a2cf0c06a6afdcafbb365ed3fb38279627/lib/RestructureCFG/SelectScopePass.cpp#L96

Reply:
Cool. Using the scope ID as key of the map you also avoid non determinism.